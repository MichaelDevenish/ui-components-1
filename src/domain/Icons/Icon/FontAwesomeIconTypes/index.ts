export { BrandIcon } from './brand'
export { RegularIcon } from './regular'
export { SolidIcon } from './solid'
