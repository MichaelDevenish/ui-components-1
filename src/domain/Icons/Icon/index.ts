export { Icon, IIconProps } from './Icon'
export {
  IconType,
  FontAwesomeIconType,
  FontAwesomeIconTypeNoPrefix,
  IntelliIconType,
  IntelliIconTypeNoPrefix
} from './types'
