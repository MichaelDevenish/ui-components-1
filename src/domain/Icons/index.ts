export * from './FontAwesomeIcon'
export * from './IntelliIcon'
export * from './Icon'
