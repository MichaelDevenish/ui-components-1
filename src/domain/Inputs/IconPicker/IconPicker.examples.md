#### Vertical Radio Set

```jsx
initialState = { value: 'option 1' };

<IconPicker
  name='icon-picker'
  value={state.value}
  handleChange={(event) => setState({value: event.target.value})}
  icons={['fa-apple', 'fa-archive', 'intelli-icon-australia', 'intelli-icon-avatar']}
/>
```
