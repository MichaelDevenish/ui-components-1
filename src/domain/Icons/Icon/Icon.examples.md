#### Basic FontAwesome Icon

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

  <FontAwesomeIcon
    type='solid'
    icon='star'
  />
```

#### Basic Intelli Icon

```jsx
import { IntelliIcon } from '@Domain/Icons';

  <IntelliIcon
    icon='smile'
  />
```

#### Extra large Icon

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

  <FontAwesomeIcon
    type='light'
    icon='star'
    size='xlarge'
  />
```

#### Duotone Icon

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';
  <FontAwesomeIcon
    type='duotone'
    icon='layer-group'
    size='xlarge'
  />
```

#### Duototone Icon with custom colors 

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

<>
  <FontAwesomeIcon
    type='duotone'
    icon='cash-register'
    size='xlarge'
    duotoneColors={{
      primaryColor: 'rgba(52, 158, 235, 1)',
      secondaryColor: '#eba313'
    }}
  />
  </>
```

#### Spinning Icon

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

  <FontAwesomeIcon
    type='regular'
    icon='star'
    isSpinning
  />
```

#### Coloured Icon

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

  <FontAwesomeIcon
    type='solid'
    icon='star'
    color='rgb(204,204,0)'
  />
```

#### Mix and match

Icon props can be used together!

```jsx
import { FontAwesomeIcon } from '@Domain/Icons';

  <FontAwesomeIcon
    type='regular'
    icon='star'
    color='rgb(204,204,0)'
    isSpinning
    size='large'
  />
```

#### Icon with badge

```jsx
import { Badge } from '@Domain/Badges';
import { FontAwesomeIcon } from '@Domain/Icons';

<div>
  <div>
    <FontAwesomeIcon
      type='light'
      icon='inbox'
      size='large'
      badge={
        <Badge
          label={6}
          backgroundColor='rgb(71, 82, 93)'
          color='white'
        />
      }
    />
  </div>

  <div>
    <FontAwesomeIcon
      type='regular'
      icon='inbox'
      size='xlarge'
      badge={
        <Badge
          label={6}
          backgroundColor='rgb(71, 82, 93)'
          color='white'
        />
      }
    />
  </div>

  <div>
    <FontAwesomeIcon
      type='solid'
      icon='inbox'
      size='xxlarge'
      badge={
        <Badge
          label={6}
          backgroundColor='rgb(71, 82, 93)'
          color='white'
        />
      }
    />
  </div>

</div>
```

#### Icon fetching

```jsx
import { Badge } from '@Domain/Badges';
import { FontAwesomeIcon } from '@Domain/Icons';

<div>
  <div>
    <FontAwesomeIcon
      type='solid'
      icon='inbox'
      size='large'
      badge={
        <Badge
          label={6}
          backgroundColor='rgb(71, 82, 93)'
          color='white'
          pending
        />
      }
    />
  </div>

  <div>
    <FontAwesomeIcon
      type='regular'
      icon='inbox'
      size='xlarge'
      badge={
        <Badge
          label={6}
          backgroundColor='rgb(71, 82, 93)'
          color='white'
          pending
        />
      }
    />
  </div>

  <div>
    <FontAwesomeIcon
      type='light'
      icon='inbox'
      size='xxlarge'
      badge={
        <Badge
          label={6}
          backgroundColor='rgb(71, 82, 93)'
          color='white'
          pending
        />
      }
    />
  </div>

</div>
```
