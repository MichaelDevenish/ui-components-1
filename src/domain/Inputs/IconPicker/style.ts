import styled, {css} from 'styled-components'

import { Variables } from '../../../common'
import { Icon } from '../../Icons/Icon'
import { GridLayout } from '../../Layouts/GridLayout'
import { Input } from '../Input'

interface IStyledIconAreaProps {
  checked: boolean
}

const StyledIconArea = styled.label<IStyledIconAreaProps>`
  width: 40px;
  height: 40px;
  border: 1px solid ${Variables.Color.n400};
  border-radius: 4px;
  margin: 0 !important;
  transition: .25s ease-out, color .25s ease-out;

  ${(props: IStyledIconAreaProps) => {
    if (props.checked) {
      return css`
        background: ${Variables.Color.i100};
        border-color: ${Variables.Color.i600};

        &:hover {
          border-color: ${Variables.Color.i400};
        }
      `
      }

      return css`
         &:hover {
          background: ${Variables.Color.n200};
        }
      `
  }}
`

const StyledIcon = styled(Icon)`
  width: 100%;
  vertical-align: middle;
`

const StyledIconInput = styled(Input)`
  height: 0;
  opacity: 0;
  position: absolute;
  width: 0;
`

export {
  StyledIconArea,
  StyledIcon,
  StyledIconInput
}
