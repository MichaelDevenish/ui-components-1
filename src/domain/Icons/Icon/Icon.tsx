import classNames from 'classnames'
import React from 'react'

import { Props } from '../../../common'
import { BadgeWrapper, StyledIcon } from './style'

interface IIconProps {
  /** FontAwesome or alternate name of the icon to display */
  icon: string
  /** Multiplies icon size by this amount */
  size?: 'xxsmall' | 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'xxlarge'
  /** Custom sizes for icons */
  customSize?: number
  /** Colour of the icon */
  color?: string
  /** Additional class name to pass to the icon */
  className?: string
  /** Adds FontAwesome `fa-spin` class to icon */
  isSpinning?: boolean
  /** Badge to display on the icon */
  badge?: JSX.Element
  /** Margins around the component */
  margins?: Props.IMargins
  /** Width of the icon */
  width?: number
  /** Shows the component only between these provided screen sizes */
  showForSizes?: Props.IShowForSizes
}

const Icon: React.FC<IIconProps> = ({
  isSpinning = false,
  size = 'medium',
  className,
  color,
  customSize,
  icon,
  badge,
  margins,
  width,
  showForSizes
}) => {
  const classes = classNames(
    className,
    'icon',
    icon,
    {
      'fa-spin': isSpinning
    }
  )

  const newBadgeSize: { [i: string]: string } = {
    large: 'small',
    xlarge: 'medium',
    xxlarge: 'large'
  }

  return (
    <React.Fragment>
      <StyledIcon
        className={classes}
        aria-hidden
        color={color}
        customSize={customSize}
        size={size}
        showForSizes={showForSizes}
        margins={margins}
        width={width}
      />
      {badge && ['large', 'xlarge', 'xxlarge'].includes(size) && (
        <BadgeWrapper
          size={size}
          showForSizes={showForSizes}
        >
          {React.cloneElement(badge, { size: size && newBadgeSize[size] })}
        </BadgeWrapper>
      )}
    </React.Fragment>
  )
}

export {
  IIconProps,
  Icon
}
