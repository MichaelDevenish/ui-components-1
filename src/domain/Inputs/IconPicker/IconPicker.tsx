import React, { ChangeEventHandler } from 'react'

import { Variables } from '../../../common'
import { StyledIcon, StyledIconArea, StyledIconInput } from './style'
import { IconType } from '../../Icons/Icon'
import { GridLayout } from '../../Layouts/GridLayout'
import { IGridLayoutCell } from '../../Layouts/GridLayout/GridLayout'

export interface IIconPickerProps {
  /** Array of icons to display in the list */
  icons: IconType[]
  /** action triggered when icons are clicked */
  handleChange?: ChangeEventHandler<HTMLInputElement>
  /** The currently selected icon */
  value?: string | number
  /** The name property of the icon picker */
  name: string
}

export class IconPicker extends React.PureComponent<IIconPickerProps> {
  public static defaultProps: Partial<IIconPickerProps> = {
    value: undefined
  }

  public render (): JSX.Element {
    return (
      <GridLayout
        gutterMarginY={Variables.Spacing.sXSmall}
        gutterMarginX={Variables.Spacing.sXSmall}
        cells={this.options}
      />
    )
  }

  private get options (): IGridLayoutCell[] {
    const {
      icons
    } = this.props

    return icons.map((icon, idx): IGridLayoutCell => {
      return {
        size: 'shrink',
        content: this.iconInput(icon, idx)
    }
    })
  }

  private iconInput = (icon: IconType, idx: number) => {
    const {
      handleChange,
      value,
      name
    } = this.props

    const checked = value === icon

    return (
      <>
        <StyledIconInput
          name={name}
          id={`${name}-${icon}-${idx}`}
          value={icon}
          handleChange={handleChange}
          isChecked={checked}
          type='radio'
        />
        <StyledIconArea checked={checked} htmlFor={`${name}-${icon}-${idx}`}>
          <StyledIcon
            type={icon}
            color={checked ? Variables.Color.i300 : Variables.Color.n700 }
            size={'large'}
          />
        </StyledIconArea>
      </>
    )
  }
}
