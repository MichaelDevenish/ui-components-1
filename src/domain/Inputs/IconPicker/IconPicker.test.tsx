import { shallow } from 'enzyme'
import React from 'react'

import { IconPicker } from './IconPicker'
import { IconType } from '../../Icons/Icon'

const dummyClick = () => console.log('hey')

const testIcons = [
  'intelli-icon-australia',
  'intelli-icon-card',
  'intelli-icon-book'
] as IconType[]

describe('<IconPicker />', () => {
  it(`should render an Icon Picker`, () => {
    const wrapper = shallow(
      <IconPicker
        name='icon-picker'
        handleChange={dummyClick}
        icons={testIcons}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
