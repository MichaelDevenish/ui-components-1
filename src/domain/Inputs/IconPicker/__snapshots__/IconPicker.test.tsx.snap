// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconPicker /> should render an Icon Picker 1`] = `
<GridLayout
  cells={
    Array [
      Object {
        "content": <UNDEFINED>
          <Styled(Input)
            handleChange={[Function]}
            id="icon-picker-intelli-icon-australia-0"
            isChecked={false}
            name="icon-picker"
            type="radio"
            value="intelli-icon-australia"
          />
          <styled.label
            checked={false}
            htmlFor="icon-picker-intelli-icon-australia-0"
          >
            <Styled(Icon)
              color="#424F5C"
              size="large"
              type="intelli-icon-australia"
            />
          </styled.label>
        </UNDEFINED>,
        "size": "shrink",
      },
      Object {
        "content": <UNDEFINED>
          <Styled(Input)
            handleChange={[Function]}
            id="icon-picker-intelli-icon-card-1"
            isChecked={false}
            name="icon-picker"
            type="radio"
            value="intelli-icon-card"
          />
          <styled.label
            checked={false}
            htmlFor="icon-picker-intelli-icon-card-1"
          >
            <Styled(Icon)
              color="#424F5C"
              size="large"
              type="intelli-icon-card"
            />
          </styled.label>
        </UNDEFINED>,
        "size": "shrink",
      },
      Object {
        "content": <UNDEFINED>
          <Styled(Input)
            handleChange={[Function]}
            id="icon-picker-intelli-icon-book-2"
            isChecked={false}
            name="icon-picker"
            type="radio"
            value="intelli-icon-book"
          />
          <styled.label
            checked={false}
            htmlFor="icon-picker-intelli-icon-book-2"
          >
            <Styled(Icon)
              color="#424F5C"
              size="large"
              type="intelli-icon-book"
            />
          </styled.label>
        </UNDEFINED>,
        "size": "shrink",
      },
    ]
  }
  gridColumns={12}
  gutterMarginX={8}
  gutterMarginY={8}
  gutterPaddingX="none"
  gutterPaddingY="none"
  horizontalAlignment="left"
  verticalAlignment="stretch"
/>
`;
